sdk/keyvault/azure-security-keyvault/tests/test_key_client.py

```diff
@@ -5,76 +5,249 @@
         if self.is_live:
-            # TODO: replace sleep with polling
-            # (client won't do it by default because 404 isn't retryable)
-            sleep(20)
-        deleted_key = client.get_deleted_key(key_name)
-        assert deleted_key.key.kid == key.id
-        keys_after_delete = client.get_all_keys()
-        key_ids_after_delete = {key.kid for key in keys_after_delete}
-        assert len(key_ids_after_delete.symmetric_difference(key_ids_at_start)) == 0
-        all_deleted_keys = client.get_all_deleted_keys()
-        recovery_ids = {key.recovery_id for key in all_deleted_keys}
-        assert delete_result.recovery_id in recovery_ids
-        if self.is_live:
-            sleep(20)
+            # wait to ensure the key has been deleted
+            time.sleep(20)
+        # get the deleted key when soft deleted enabled
+        deleted_key = client.get_deleted_key(created_rsa_key.name)
+        self.assertIsNotNone(deleted_key)
+        self.assertEqual(created_rsa_key.id, deleted_key.id)
 
-        # recover the deleted key
-        recovered_key = client.recover_deleted_key(key_name)
-        assert recovered_key.id == key.id
+    @ResourceGroupPreparer()
+    @VaultClientPreparer()
+    def test_backup_restore(self, vault_client, **kwargs):
+
+        self.assertIsNotNone(vault_client)
+        client = vault_client.keys
+        key_name = self.get_resource_name("keybak")
+        key_type = "RSA"
+
+        # create key
+        created_bundle = client.create_key(key_name, key_type)
+        self.assertEqual(key_type, created_bundle.key_material.kty)
+
+        # backup key
+        key_backup = client.backup_key(created_bundle.name)
+        self.assertIsNotNone(key_backup, "key_backup")
+
+        # delete key
+        client.delete_key(created_bundle.name)
+
+        # restore key
+        restored = client.restore_key(key_backup)
+        self._assert_key_attributes_equal(created_bundle, restored)
+
+    @ResourceGroupPreparer()
+    @VaultClientPreparer()
+    def test_key_list(self, vault_client, **kwargs):
+
+        self.assertIsNotNone(vault_client)
+        client = vault_client.keys
+
+        max_keys = self.list_test_size
+        expected = {}
+
+        # create many keys
+        for x in range(0, max_keys):
+            key_name = "key{}".format(x)
+            key = client.create_key(key_name, "RSA")
+            expected[key.name] = key
+
+        # list keys
+        result = list(client.list_keys(max_page_size=max_keys))
+        self._validate_key_list(result, expected)
+
+    @ResourceGroupPreparer()
+    @VaultClientPreparer()
+    def test_list_versions(self, vault_client, **kwargs):
+
+        self.assertIsNotNone(vault_client)
+        client = vault_client.keys
+        key_name = self.get_resource_name("testKey")
+
+        max_keys = self.list_test_size
+        max_page_size = 2
+        expected = {}
+
+        # create many key versions
+        for _ in range(0, max_keys):
```

Comment:
FYI given one argument, `range` treats that argument as the upper bound, and defaults to a lower bound of `0`, i.e. this is equivalent to `range(max_keys)`.